src/sctools/groups.py

```diff
@@ -0,0 +1,151 @@
+        # results to include TANDEM repeats. but we only output the first line.
+        elif class_name == "InsertSizeMetrics":
+            # if the elemnet counts is less than 21,
+            # it means insertion metrics returns multiple line results.
+            if len(contents) < 21:
+                rows = contents[0]
+            else:
+                rows = contents
+        else:
+            # other metrics(so far) only return one line results.
+            rows = contents
+        metrics[cell_id].update({
+                k: rows[k] for k in rows if k not in
+                ['SAMPLE', 'LIBRARY', 'READ_GROUP', 'CATEGORY']
+                })
+        df = pd.DataFrame.from_dict(metrics, orient='columns')
+        df.insert(0, 'Class', class_name)
+        d = d.append(df)
+    d.to_csv(output_name + '.csv')
+
+
+def aggregate_picard_metrics_by_table(file_names, output_name):
+    """
+    pipeline output picard QC metrics at single cell/sample level.
+    This function is called to merge/aggregate QC metrics by metrics type and
+    then merge multiple QC measurement into single matrix file.
+    In this file, column is sample/cell and row is QC metrics
+    :param file_names: list of files generated by Picard, metrics are in table.
+    :param output_name: output name without suffix'
+   """
+    for file_name in file_names:
+        cell_id = os.path.basename(file_name).split('_qc')[0]
+        class_name = os.path.basename(file_name).split('.')[1]
+        parsed = picard.parse(file_name)
+        dat = pd.DataFrame.from_dict(parsed['metrics']['contents'])
+        dat.insert(0, 'Sample', cell_id)
+        dat.to_csv(output_name + "_" + class_name + '.csv', index=False)
+
+
+def aggregate_qc_metrics(file_names, output_name):
+    """
+    Append several metrics into one file. 
+    :param file_names: list of files, such as Picard row metric, hisat2 metric
+    and rsem metric
+    :param output_name: output file name without suffix.
+    """
+    df = pd.DataFrame()
+    for file_name in file_names:
+        dat = pd.read_csv(file_name)
+        df = df.append(dat, ignore_index=True)
+    df.rename(columns={'Unnamed: 0': 'Metrics'}, inplace=True)
+    df.to_csv(output_name + '.csv', index=False)
+
+
+def parse_hisat2_log(file_names, output_name):
+    """
+    to parse the hisat2 log file, which include alignment summary
+    :param file_names: hisat2 log files
+    :param output_name: output file name without suffix
+    """
+    metrics = {}
+    tag = "NONE"
+    for file_name in file_names:
+        if '_qc' in file_name:
+            cell_id = os.path.basename(file_name).split('_qc')[0]
+            tag = "HISAT2G"
+        elif '_rsem' in file_name:
+            cell_id = os.path.basename(file_name).split('_rsem')[0]
+            tag = "HISAT2T"
+        with open(file_name) as f:
+            dat = f.readlines()
+            d = [x.strip().split(':') for x in dat]
+            d.pop(0)
```

Comment:
Please add some comments about these magic numbers. What's going on with the pop, splitting?

Reply:
it will pop out the first row of each section.

Reply:
Cool. Why are you doing this? Is it a header that you don't need? (put it into the code comment, please)